### PR TITLE
adding debug-dump-gcs to dump hlo and upload to the gcs bucket

### PR DESCRIPTION
**Pull request by ssusie**
The flag sets the  XLA_FLAGS to dump to /tmp/xla_dump dir then uploads to provided GCS bucket. This however may overwrite the XLA_FLAGS passed by user through env file

### CODE REVIEW

File: xpk/xpk.py [hunk 1]
```diff
@@ -473,6 +473,13 @@ def add_env_config(args):
               'environment, a value must be specified.'
           )
           env[variable] = os.environ[variable]
+
+  if args.debug_dump_gcs:
+    if 'XLA_FLAGS' in env:
+      # Caveat: might be overwriting something passed by user
+      print('overwriting XLA_FLAG from env-file')
```
Comment (Obliviour) [inline xpk/xpk.py:480]: Nit 'xpk_print'

File: xpk/xpk.py [hunk 2]
```diff
@@ -1998,6 +2010,14 @@ def directory_path_type(value):
         " Defaults to 0."
     ),
 )
+workload_create_parser_optional_arguments.add_argument(
+    '--debug-dump-gcs',
+    type=str,
+    default=None,
+    help=(
+        "gcs bucket where debugging info like hlo dump is uploaded to"
```
Comment (Obliviour) [inline xpk/xpk.py:2018]: Nit to add an example of an argument to help guide the user. Also nit to add period. Ty!

File: xpk/xpk.py [hunk 3]
```diff
@@ -1455,6 +1462,11 @@ def workload_create(args) -> int:
     xpk_exit(setup_docker_image_code)
 
   add_env_config(args)
+  if args.debug_dump_gcs:
+    args.command += ('; WORKER_ID=$(python3 -c "import jax; print(jax.process_index())");'
```
Comment (Obliviour) [inline xpk/xpk.py:1466]: Curious if the process_index print is needed

Comment (rwitten) [reply xpk/xpk.py:1466]: The print here is being used to capture the process index into args.command. Interesting idea.

Two thoughts:
* args should be immutable (so unless this is inside of initializing args we need a new variable and stitch them togeter)
* this code will run jax.process_index() which if I recall correctly retriggers slice builder and MXLA. That is probably pretty unexpected to the user. Obviously you do want an identified but how about $HOSTNAME?

**Review by rwitten — CHANGES_REQUESTED** [review]
Little nits but broadly good!

File: xpk/xpk.py (hunk 1 above)
Comment (ssusie) [reply xpk/xpk.py]: Thanks, changed to raising an error as per Rafi's suggestion

File: xpk/xpk.py (hunk 2 above)
Comment (ssusie) [reply xpk/xpk.py]: Done, thanks.

File: xpk/xpk.py (hunk 3 above)
Comment (ssusie) [reply xpk/xpk.py]: Modified and used the full $HOSTNAME, thanks for the suggestion!

File: xpk/xpk.py [hunk 4]
```diff
@@ -473,6 +473,13 @@ def add_env_config(args):
               'environment, a value must be specified.'
           )
           env[variable] = os.environ[variable]
+
+  if args.debug_dump_gcs:
+    if 'XLA_FLAGS' in env:
+      # Caveat: might be overwriting something passed by user
+      print('overwriting XLA_FLAG from env-file')
+    env['XLA_FLAGS'] = '--xla_dump_to=/tmp/xla_dump/'
```
[the remark this answers is not in the document]
Comment (ssusie) [reply xpk/xpk.py:482]: Great point, modified to raise an error.

File: xpk/xpk.py [hunk 5]
```diff
@@ -473,6 +473,14 @@ def add_env_config(args):
               'environment, a value must be specified.'
           )
           env[variable] = os.environ[variable]
+
+  if args.debug_dump_gcs:
+    if 'XLA_FLAGS' in env:
```
Comment (rwitten) [inline xpk/xpk.py:478]: Shouldn't we check if `xla_dump_to` is in the XLA_FLAGS? That way the user can set `XLA_FLAGS` (for other things).

File: xpk/xpk.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rwitten) [reply xpk/xpk.py:482]: This is now saying the user can't use XLA_FLAGS which seems wrong. We want them to be able to use `XLA_FLAGS` but not `xla_dump_to`, right?

**Review by rwitten — CHANGES_REQUESTED** [review]

**Review by rwitten — APPROVED** [review]

File: xpk/xpk.py (hunk 5 above)
Comment (Obliviour) [reply xpk/xpk.py:483]: +1 we can support more adding `xla_dump_to` the XLA_FLAGS if `xla_dump_to` is not specified by the env file.

**Review by Obliviour — APPROVED** [review]

**Pull request merged**